src/realm/sync/network/default_socket.cpp

```diff
@@ -65,25 +65,84 @@ class DefaultWebSocketImpl final : public WebSocketInterface, public Config {
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::ReadError, ec.message()});
     }
     void websocket_write_error_handler(std::error_code ec) override
     {
         m_logger.error("Writing failed: %1", ec.message()); // Throws
-        m_observer.websocket_read_or_write_error_handler(ec);
+        constexpr bool was_clean = false;
+        websocket_error_and_close_handler(was_clean, Status{ErrorCodes::WriteError, ec.message()});
     }
     void websocket_handshake_error_handler(std::error_code ec, const HTTPHeaders*,
                                            const std::string_view* body) override
     {
-        m_observer.websocket_handshake_error_handler(ec, body);
+        ErrorCodes::Error error;
+
+        if (ec == websocket::Error::bad_response_3xx_redirection ||
+            ec == websocket::Error::bad_response_301_moved_permanently) {
+            error = ErrorCodes::WebSocket_MovedPermanently;
+        }
+        else if (ec == websocket::Error::bad_response_401_unauthorized) {
+            error = ErrorCodes::WebSocket_Unauthorized;
+        }
+        else if (ec == websocket::Error::bad_response_403_forbidden) {
+            error = ErrorCodes::WebSocket_Forbidden;
+        }
+        else if (ec == websocket::Error::bad_response_5xx_server_error ||
+                 ec == websocket::Error::bad_response_500_internal_server_error ||
+                 ec == websocket::Error::bad_response_502_bad_gateway ||
+                 ec == websocket::Error::bad_response_503_service_unavailable ||
+                 ec == websocket::Error::bad_response_504_gateway_timeout) {
+            error = ErrorCodes::WebSocket_5xx_Server_Errors;
+        }
+        else {
+            error = ErrorCodes::WebSocket_Protocol_Mismatch;
+            if (body) {
+                std::string_view identifier = "REALM_SYNC_PROTOCOL_MISMATCH";
+                auto i = body->find(identifier);
+                if (i != std::string_view::npos) {
+                    std::string_view rest = body->substr(i + identifier.size());
+                    // FIXME: Use std::string_view::begins_with() in C++20.
+                    auto begins_with = [](std::string_view string, std::string_view prefix) {
+                        return (string.size() >= prefix.size() &&
+                                std::equal(string.data(), string.data() + prefix.size(), prefix.data()));
+                    };
+                    if (begins_with(rest, ":CLIENT_TOO_OLD")) {
+                        error = ErrorCodes::WebSocket_Client_Too_Old;
+                    }
+                    else if (begins_with(rest, ":CLIENT_TOO_NEW")) {
+                        error = ErrorCodes::WebSocket_Client_Too_Old;
+                    }
+                }
+            }
+        }
+
+        constexpr bool was_clean = false;
```

Comment:
Except for `ErrorCodes::WebSocket_InternalServerError`, `was_clean` should be `true`, since these will eventually be sent by the server.